lib/BuildSystem/LaneBasedExecutionQueue.cpp

```diff
@@ -126,14 +152,42 @@ class LaneBasedExecutionQueue : public BuildExecutionQueue {
     for (unsigned i = 0; i != numLanes; ++i) {
       lanes[i]->join();
     }
+
+    if (killAfterTimeoutThread) {
+      stopKillingCondition.notify_all();
+      killAfterTimeoutThread->join();
+    }
   }
 
   virtual void addJob(QueueJob job) override {
+    if (cancelled && !job.isSentinel()) {
```

Comment:
What I would do instead of checking `job.isSentinel()` here, which seems rather subtle, is to `notify_all()` in `cancelAllJobs()` to ensure that any that are blocked awake in order to see the cancellation flag have become true.